### DISCUSSION TITLE
Vendure and Cookies

### DISCUSSION THREAD

**Discussion by rohanrajpal** [vendurehq/vendure#638] [Q&A] [answered]
Hey, I had an idea to make a thread for anyone who is having trouble with setting up cookies with Vendure.

@michaelbromley shared a pretty good GitHub gist which helps in setting up Nginx the right way to allow cookies
[1]: https://gist.github.com/nikmartin/5902176

I also found the following repo quite helpful in debugging cookie issues
https://github.com/benawad/how-to-debug-cookies#proxy

I also had a few questions

- if we are following [1], is the following cookie configuration fine?
```
cookieOptions: {
      secure: true,
      domain: ".example.com",
    },
```

- I tried to add `app.set("trust proxy", 1);` in the BeforeVendureBootStrap plugin, but came across the following issues:

```ts
// plugin.ts
import { INestApplication } from '@nestjs/common';
import { VendurePlugin } from '@vendure/core';

@VendurePlugin({})
export class ProxyPlugin {
  static beforeVendureBootstrap(app: INestApplication): void | Promise<void> {
    app.set("trust proxy", 1);
    //^^ Property 'set' does not exist on type 'INestApplication'.ts(2339)
  }
}
```
So what is the correct way to `app.set("trust proxy", 1);`?

**Comment by michaelbromley** [comment, marked as answer]
Regarding the `app.set()` issue, you can cast the `app` object to be `NestExpressApplication` (which is what it actually is):

```
(app as NestExpressApplication).set(...);
```

Reference: https://github.com/nestjs/nest/issues/2293#issuecomment-500020645

**Reply by rohanrajpal** [reply]
Thanks, Michael, this worked. I had another question, what do I have to do in Vendure to achieve the following configuration?

```ts
// 1. In your main App, setup up sessions:

app.enable('trust proxy'); //done via plugin
app.use(express.bodyParser()); // assuming vendure handles this
app.use(express.cookieParser()); // assuming vendure handles this
app.use(express.session({
   secret: 'Super Secret Password', // handled with the vendure config
   proxy: true, // how do I enable this?
   key: 'session.sid', // is this necessary to set?
   cookie: {secure: true}, // I guess this is same as the `secure` field in `cookieOptions`?
   store: new sessionStore()  // I guess we can ignore this as well?
}));
```

**Reply by rohanrajpal** [reply]
I think to enable the proxy I have to use a plugin as well. but the issue is where should I get the express variable from?

**Reply by michaelbromley** [reply]
Vendure internally uses the [cookie-session middleware](https://expressjs.com/en/resources/middleware/cookie-session.html), you can see the source here: https://github.com/vendure-ecommerce/vendure/blob/dc7b3036e5ade97c733cd6b42b04749d49ada294/packages/core/src/bootstrap.ts#L58-L67

I'm not 100% sure about the relationship between cookieSession and cookieParser and session, i.e. whether "cookieSession" is a replacement for the other two. Sorry I can't help more on this, I have yet to dive deep into these issues myself.

Also bodyParser is applied automatically by Nestjs.

**Reply by rohanrajpal** [reply]
Oh I understand now. I just guess setting `proxy:true` is left in the config part, do you know anyone who could help me with that?

**Reply by rohanrajpal** [reply]
Forgot to reply, but I resolved the cookie issue! For anyone facing any issues, the above answer solves it.

**Comment by knoid** [comment]
Is there an updated answer to this? `beforeVendureBootstrap` doesn't seem to be working anymore.

**Reply by michaelbromley** [reply]
Hi @knoid, that hook was removed because it turned out you can use the built-in NestJS module lifecycle to do the same things.

For this case you should be able to inject the NestJS `HttpAdapterHost` into the `onApplicationBootstrap` hook of your plugin module.
